### ISSUE TITLE
GFX Performance degradation inquiry

### ISSUE THREAD

**Issue by freerdp-github** [FreeRDP/FreeRDP#3846]
Hoping one of the devs can help me find where to begin looking for why our GFX performance went from good to slow in xfreerdp. The good version (2.0.0-dev (git dbb353d)), built from master on 10/03/16, was performing very well using the command line:
$BIN/xfreerdp /gfx +gfx-progressive -bitmap-cache -offscreen-cache -glyph-cache +clipboard -themes -wallpaper /f /u:$username /p:$password /d:DOMAIN /sec:nla /cert-ignore /v:${ips[$cnt]} 
on an Ubuntu 14.04.5 Trusty thin client. The latest version (2.0.0-dev (git 5bb7a05)) built from the most recent master using the same command line is noticeably slower. The thin client is connecting to a Windows 2012 R2 RDS server and running a popular CAD software package. The performance degradation is most noticeable during model rotations and simulations where excessive tearing and lagging is apparent. I realize this is a six month window between commits but I was hoping one of the devs working on the GFX code could pinpoint an area to start looking at so we can get back to a certain performance level without regressing.

Here is the build config (both versions use the identical one):

$ /usr/local/freerdp/bin/xfreerdp /buildconfig
This is FreeRDP version 2.0.0-dev (git dbb353d)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EVENTFD_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_INTTYPES_H=1 HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CHANNELS=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=OFF WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_GDI=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_ORDERS=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DIRECTFB=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FFMPEG=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_IPP=OFF WITH_JPEG=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=OFF WITH_PROFILER=OFF WITH_PULSE=OFF WITH_SAMPLE=OFF WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SSE2=ON WITH_THIRD_PARTY=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_WAYLAND=OFF WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          Debug
CFLAGS:               -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -msse2 -DWINPR_EXPORTS -DFREERDP_EXPORTS
Compiler:            GNU, 4.8.4
Target architecture: x64

**Comment by akallabeth** [comment]
you use build type debug, which basically disables every known optimisation of the compiler. could that be the reason?

**Comment by akallabeth** [comment]
@freerdp-github Oh, just to clarify, we've added a lot of code checks that slow down non optimized compiles but help during development. That in combination with your debug build can indeed be a performance issue.

**Comment by freerdp-github** [comment]
@akallabeth Thanks for your help. This is really fantastic software.
Using the "Release" build for the most recent FreeRDP master has better performance
then what we were seeing with the "Debug" build. However, there does appear to still be
some performance loss during a CAD verify operation. The verify operation depicts a tool
cutting the surface of a part in realtime. The graphic representing the tool breaks up 
and lags behind the actual cutting operation, catching up eventually. Sorry I cannot 
show you this. This is the only difference in performance as all other operations (rotations,
etc) seem to be acceptable as before.

Background: We have been using the "/gfx +gfx-progressive" command line options with success.
Our thin clients are based on the AMD Embedded GX-415GA SOC (Kabini) with AMD Radeon HD 
8330E Graphics, so I compiled and use the proprietary fglrx AMD driver in Ubuntu 14.04.5 until I can 
migrate to the open source radeon/amdgpu drivers in 16.04.2.

1. Would the openh264 codec provide better CAD performance? "/gfx:AVC420 /gfx-h264:AVC420"
2. Any other ideas as to what codecs/options could provide better performance?
3. Is there any documentation about the various graphics command line options?

Thanks again

**Comment by akallabeth** [comment]
@freerdp-github Thank you :)
As for your CAD issues, is this operation sending many updates to the client? (Network bandwidth usage should show you increased traffic)
Otherwise it is most likely the server is not fast enough with encoding and you should optimize on that end.
RFX/GFX/AVC modes are basically doing all encoding server side and the client must only be sufficiently fast to decode the video stream.
As for AVC modes, AVC420 is quite buggy, not really usable when a good contrast is necessary (hello CAD) and only supported in winows 8 and newer.
AVC444 on the other hand has good contrast but is only supported with Windows 10 (update someting, I forgot) so your windows 2012 servers will not support that.

**Comment by akallabeth** [comment]
@freerdp-github Oh, and as a reference, decoding on a raspberry pi 3 is fast enough to get ~15 FPS (FreeRDP is software decoding only so far)

**Comment by freerdp-github** [comment]
We have high speed fiber network backbone and the RDS servers are high-end. The old xfreerdp version we were using that performed better then the newest master was even compiled as a Debug build so something has changed with the newer code base if we cannot get back to the same level of performance with the latest master built as a Release build. I am going to put the old version image back on the test thin client to re-examine the performance. If I remember correctly, just before I rebuilt FreeRDP I did an apt update/upgrade on 14.04.5 which upgraded gstreamer. When I rebuilt FreeRDP, the script of the build only showed one compile change:

[ 81%] Building C object channels/tsmf/client/gstreamer/CMakeFiles/tsmf-client-gstreamer-decoder.dir/tsmf_gstreamer.c.o
[ 81%] Building C object channels/tsmf/client/gstreamer/CMakeFiles/tsmf-client-gstreamer-decoder.dir/tsmf_X11.c.o

I don't think this matters but I thought I would mention it. And yes, I noticed Windows 2016 Server is supporting the AVC444 codec. I will keep monitoring and hopefully we can find the bottleneck.

**Comment by akallabeth** [comment]
@freerdp-github Interesting. Do you know if the processor utilisation of the thin clients? The only lag I can reproduce in my test setup are server related, but maybe I'm missing something.

**Comment by bmiklautz** [comment]
@freerdp-bot can you provide us a cachegrind/callgrind trace file?
```
valgrind --tool=callgrind xfreerdp ...
valgrind --tool=cachegrind xfreerdp ...
```
For the traces it is also important to have the exact git version you are using along with some information on the hardware (including graphic cards).

**Comment by joakim-tjernlund** [comment]
@freerdp-github You could use git bisect to find the offending commit

**Comment by freerdp-github** [comment]
During the FreeRDP upgrade, many other changes occurred; including: FreeRDP update, Ubuntu update, CAD app upgrades, nVidia driver updates on server among other changes. I have reverted the thin client image to its original state so we are at least back to the same good performance prior to the updating. That eliminates server updates, both CAD app and driver changes from this problem. I have now introduced the upgraded FreeRDP (latest master) once again with out the Ubuntu updates to see how that responds. The Ubuntu upgrade updated 148 packages including codec, av, gstreamer related packages so I want to eliminate the possibility that Linux changes introduced problems. This test will tell me definitively if the FreeRDP changes are the problem. If it is FreeRDP, perhaps I may embark on the "bisect" idea, although 6 months of commits is a lot of testing. If it is Linux, I have made a comprehensive list of packages changed, so this may be helpful. I will get back to you after more testing.

**Comment by joakim-tjernlund** [comment]
You will be surprised how few rebuilds bisect will need. I took me about 12 rebuilds to
find a regression between kernel 4.6 and 4.7

**Comment by akallabeth** [comment]
any news on this one?

**Comment by freerdp-github** [comment]
Sorry for the delay; been trying to eliminate other updates from this scenario. I can now rule out Linux updates, Windows updates, nvidia driver updates, and CAD application updates from this problem. Therefore, 100% sure FreeRDP has changed and is affecting the peformance. We rebuilt the older version as a "Release" build and the performance is excellent so we are currently setting up an environment in which to reproduce the problem in a predictable way in order to quickly bisect, build, and test. If created will also generate cachegrind/callgrind trace files. Hopefully, we can get to the proper regression quickly.

Do any of the devs know if not having the freerdp ld cache updated/created properly can affect performance? One of our newer builds might not of had ldconfig updated after the build or had the conf file commented out. We are testing this as well. If we cannot turn the results around in a timely manner, we will ask to close this thread and comment later.

**Comment by freerdp-github** [comment]
I have generated call/cachegrind files during a session that I witnessed the tearing and 
laggy behaviour. I attached images showing the tool before and after tearing to give you 
some idea of what we are seeing. We also checked whether the ldconfig cache issues came into
play and it did not. Bisects will have to wait for now, hopefully  the valgrind files show
what is happening.

[gfx-issues-valgrind-files.zip](https://github.com/FreeRDP/FreeRDP/files/855691/gfx-issues-valgrind-files.zip)

![1500-test-part-broken-tool-artifacts-cropped](https://cloud.githubusercontent.com/assets/16350618/24110741/55e90e0c-0d38-11e7-9d34-a3498e29b28e.png)

![1500-test-part-good-tool-picture-cropped](https://cloud.githubusercontent.com/assets/16350618/24110742/55f21f88-0d38-11e7-9efe-ba5f1f08bff0.png)

Thanks

**Comment by akallabeth** [comment]
@freerdp-github Thanks for that info, just one question: At which resolution do you run the sessions? Did not find that information anywhere.

**Comment by akallabeth** [comment]
@freerdp-github Another hint, ```progressive``` codec is much less optimized than default ```RFX```, is there any specific reason you use that flag?

**Comment by freerdp-github** [comment]
Sorry about that; the resolution used during these tests were at 1920x1080 on a 22" monitor. I attached the output of xdpyinfo and the log from Xorg for a complete summary of Xorg capabilities on the thin client.

[xdpyinfo.out.txt](https://github.com/FreeRDP/FreeRDP/files/859946/xdpyinfo.out.txt)
[Xorg.0.log.txt](https://github.com/FreeRDP/FreeRDP/files/859947/Xorg.0.log.txt)

The reason we are using the progressive codec is from empirical work performed as a result of the #3030 thread we started in December/2015. At that time we were using both the /rfx and /gfx options together. However, this was causing random disconnects between the client and server. The only way to alleviate this problem was to drop the /rfx switch. The /gfx codec performed so well that in all honesty we never thought of going back and testing with just the /rfx switch.

Given the following successful command line from that earlier thread:

$BIN/xfreerdp /gfx +gfx-progressive -bitmap-cache -offscreen-cache -glyph-cache -themes -wallpaper /d:DOMAIN /sec:nla /cert-ignore /f +auto-reconnect /u:$USER /p:$PASS /v:$IP

What would you recommend trying with RFX in it's place? We will test this as well.

The issues however with GFX still raises questions; what has happened to the code to slow down the performance so dramatically. I have tried to eliminate the local changes that could affect performance. 

Appreciate your help in this matter....

**Comment by akallabeth** [comment]
@freerdp-github Strange, ```/rfx``` and ```/gfx``` are just flags, if available ```/gfx``` is used and ```/rfx``` just a fallback if the first is not available. (```/rfx``` is the old windows 7 style, ```/gfx``` windows 8 and newer redirection, both use the same codec).
Anyway, as long as your CPU utilisation is nowhere near 100% the codec is completely irrelevant for your case. (Decoding is done software only)
Your issues are with drawing, which is the same regardless of codec.
The only switch influencing that is ```/gdi:sw``` or ```/gdi:hw```.
I'd also add ```+fonts``` to your command line.

**Comment by joakim-tjernlund** [comment]
I just remembered, a few weeks ago I did a demo over MS Lync(using Pidgin) where I also shared my screen.
On the receiving end they recorded this demo and I noticed that my screen share was seconds late
w.r.t audio. I wonder if this lag is related to this issue?

**Comment by joakim-tjernlund** [comment]
Fond these two links about sharing:
 https://blogs.technet.microsoft.com/scottstu/2014/08/20/an-update-on-desktop-sharing-in-lync/
and 
 http://blog.schertz.name/2015/10/video-based-screen-sharing-in-skype-for-business/

Does freerdp adjust to higher FPS automatically?

**Comment by akallabeth** [comment]
@joakim-tjernlund We just found 2 regressions with #3883 and #3882 which negatively influence the RFX and GFX performance of xfreerdp. With these two the average performance should be back up.

**Comment by akallabeth** [comment]
@joakim-tjernlund And yes, FPS are adjusted according to the resources available on the client (simplified that is the client sends a Frame Acknowledge to the server, which adjusts the FPS accordingly)

**Comment by joakim-tjernlund** [comment]
NIce, these two regressions also affected perf when I share to someone else/conf. room?

**Comment by akallabeth** [comment]
@joakim-tjernlund does not matter, RDP is RDP no matter how it is wrapped.

**Comment by joakim-tjernlund** [comment]
hmm, you mean that direction does not matter w.r.t this regression?
Me sharing my screen to someone else or me looking at some other remote screen, I should
see improvements in both cases?

**Comment by nfedera** [comment]
@joakim-tjernlund just build the current master and try. @freerdp-github please do the same.

**Comment by joakim-tjernlund** [comment]
Already built, but will be some time before real test over high latency link

**Comment by joakim-tjernlund** [comment]
@akallabeth Even when doing Lync/S4B P2P sharing I don't get more than c.a 2 FPS. I guess the server is limiting this too. Is there way for me to ignore/increase the FPS limit the server artificially imposes?
I don't mind a crude C hack to try it out

**Comment by akallabeth** [comment]
You can play with ```/frame-ack:<number>``` and see if that helps

**Comment by freerdp-github** [comment]
Unfortunately, after all morning testing with the latest master, the users are saying not only is the performance worse than before (during the Verify simulation) but now they see the tearing artifacts show up in the main application window when doing graphical operations there (similar to Verify just not as elaborate). We have gone back to the earlier version  (2.0.0-dev (git dbb353d)) for the time being.

**Comment by akallabeth** [comment]
@freerdp-bot Ok, please clarify performance please.
I've prof (performance test runs) that the drawing routines are now significantly faster. 
Are you talking only about the tearing?
Trying to get the root of you issue, but it looks like there is some misunderstanding in vocabulary here.

**Comment by akallabeth** [comment]
@freerdp-github And just for reference, could you test your setup with the open source AMD drivers?

**Comment by nfedera** [comment]
@freerdp-github what is the name of that "popular CAD software" and is there a demo version available?

**Comment by joakim-tjernlund** [comment]
One thing I have noted when screen sharing via Lync is that the FPS feels a lot lower than 2.5 FPS
and the Lync server is supposed to enforce a 2.5 FPS. This slowness is also visible when sharing
over a local GBE LAN. it almost feels like freerdp miscalculates the allowed FPS.

**Comment by akallabeth** [comment]
@joakim-tjernlund Ok, reread your comments, you're using FreeRDP-shadow to encode your local screen and share it with someone else, correct? The fixes above just improve performance if FreeRDP is used as client (```xfreerdp```)

**Comment by joakim-tjernlund** [comment]
Ok, I hope a Remmina client also benefits from the fixes?

**Comment by akallabeth** [comment]
@joakim-tjernlund No, Remmina uses its own drawing. They are xfreerdp only.

**Comment by freerdp-github** [comment]
@akallabeth performance issues remain the same as previously stated, tearing of the tool, the tool lagging or stuttering behind, and eventually catching up with a jerky motion. The normal motion of the tool when correct should be consistent, unobstructed, deliberate, and free flowing as before. The open source AMD drivers are only available in the new Xorg series starting in Ubuntu 16.04. We tried this Linux OS version when it was first released on our test thin client and actually was very impressed with it in general. The performance of FreeRDP at that time was as good or even a little better then the closed source fglrx driver from AMD. Of course the xfreerdp version used was even earlier than 2.0.0-dev (git dbb353d)) that is used today. Unfortunately, Ubuntu 16.04 has several well documented problems (suspend to ram issues, drm driver problems, stability issues) that have precluded using it. Now that the 16.04.2 point release has been available we plan on giving it another try when tinker time is available.

@nfedera the popular CAD package is called "Mastercam 2017". You would use the "Verify" simulation feature of the software to recreate this phenomenon. I believe you can download a fully
functional Home learning version [here](http://www.inhousesolutions.com/mastercam-2017-hle-download/)

Unfortunately, I am not onsite at this company any longer and therefore will not be able to test "hands-on" for quite some time, sorry. Good luck!

**Comment by akallabeth** [comment]
@freerdp-github Ok, that clarifies the situation. @nfedera has done some analysis in that regard and came up with #3891 and I've adjusted it for ```/gdi:hw``` with #3892 
They both limit the update region to changed rectangles instead of updating the whole bounding rectangle of the screen improving average performance back (and beyond) to previous state.

**Comment by nfedera** [comment]
@freerdp-github thanks for the update. I will test the mastercam 2107 demo here.

**Comment by nfedera** [comment]
@freerdp-github I don't really have the time to learn mastering this complex cam software. Can you provide a quick instruction how to get "something moving" in mastercam?
Which file should I load and do I initiate the simulation ?

**Comment by freerdp-github** [comment]
@akallabeth did you mean you adjusted it for /gdi:sw not hw? Isn't /gdi:sw the default? We don't use the /gdi:hw switch.

@nfedera no problem, let me put a test part file together for you with instructions to get you going.

**Comment by akallabeth** [comment]
@freerdp-github It is fixed for both implementations now. Just a note, ```/gdi:hw``` is the current default.

**Comment by nfedera** [comment]
@freerdp-github thanks for the detailed instructions. However, when clicking "Verify selected operations" button in the Toolpaths area I'm getting the following error:
![image](https://cloud.githubusercontent.com/assets/670606/24601164/2435dd6a-1858-11e7-8b25-b02509ab6440.png)

But the "c:\program files\McamDemoHLE2017\chooks\MachSim.dll" is present.

**Comment by nfedera** [comment]
@freerdp-github never mind. The error is actually misleading. MachSim.DLL is linked against a Mediafoundation dll which is missing on Server 2012R2 unless that feature is activatedin the server manager.

**Comment by nfedera** [comment]
@freerdp-github with the most recent changes/fixes I cannot see any performance difference (compared to the old version) anymore.

Here are three videos comparing your mastercam slowmo simulation demo.
- [using Microsoft mstsc.exe on a Windows 7 Workstation](https://vid.me/FcxX)
- [using FreeRDP from Sep 2016 (dbb353d) on a Raspberry Pi 2 B](https://vid.me/G8vk)
- [using FreeRDP current master (d492d52) on a Raspberry Pi 2 B](https://vid.me/EXm1)

**Comment by freerdp-github** [comment]
@nfedera thanks for taking the time to do the videos. I built and installed the latest master with the fixes (d492d52) and we have been testing for two days now and I think you are on the right track. The performance is much better but there is still an occasional glitch happening that most people probably would not even notice it is so fast. In all your testing situations (videos) you show the whole screen. When this tool is used in real situations, the user normally has the cylinder and tools zoomed in to see the cutting action. Looking at the screen shots posted earlier in the thread show just about how close this zoom is. When running the simulation at this zoom level you can sometimes see these issues. Again, it is much better now with pull #3891 so I think you have nailed most of the problem.

If you want to test the zoom later then simply right-click over the model and select "Fit" to align model. Then right-click again and select "Zoom Window", left-click the mouse and hold down to draw a bounding box around the cylinder and top of tool and then left click. Do this until the placement of the model is similar to the above images (you may also use PG-UP/DN). Then run the simulation with the current master to see if you see any issues. We found that on the test part file we no longer could see any issues, however with a real part file we noticed only slight issues. Again, barely perceptible unless you knew what you were looking for. So all in all I would say you are close. We are using the new version and the user appears to be happy with it. I am not sure if some of the differences we are seeing is because of the black background or the zooming effect but at least you have another good app to test with now. I would think the black background would be easier to draw then the gradient you are using so I don't think it is that. So the zooming could do it because it is a larger rectangle that is changing causing more redraws. Also, on the Windows 7 PC, were you using RDP7 or 8? Thanks for your effort with this. Let me know if you need anything else.

**Comment by freerdp-github** [comment]
@akallabeth We finally got around to installing the latest point release Lubuntu 16.04.2 with the HWE updated kernel and Xorg open source drivers for the Radeon chipset. The newer 4.8 series kernels have fixed the issues that precluded us from using it on our thin clients. The latest FreeRDP master runs very well on this new setup so I am going to close this out now. I want to thank you and @nfedera especially for your support.
